lib/operations/common/GitlabRepoRef.ts

```diff
@@ -0,0 +1,157 @@
+import * as url from "url";
+import {
+    ActionResult,
+    successOn,
+} from "../../action/ActionResult";
+import { Configurable } from "../../project/git/Configurable";
+import { DefaultHttpClientFactory } from "../../spi/http/axiosHttpClient";
+import { HttpMethod } from "../../spi/http/httpClient";
+import { logger } from "../../util/logger";
+import { AbstractRemoteRepoRef } from "./AbstractRemoteRepoRef";
+import { GitlabPrivateTokenCredentials } from "./GitlabPrivateTokenCredentials";
+import { GitShaRegExp } from "./params/validationPatterns";
+import { ProjectOperationCredentials } from "./ProjectOperationCredentials";
+import { ProviderType } from "./RepoId";
+
+export const GitlabDotComApiBase = "https://gitlab.com/api/v4";
+export const GitlabDotComRemoteUrl = "https://gitlab.com/";
+
+/**
+ * Repository reference implementation for Gitlab
+ */
+export class GitlabRepoRef extends AbstractRemoteRepoRef {
+
+    public static from(params: {
+        owner: string,
+        repo: string,
+        sha?: string,
+        rawApiBase?: string,
+        path?: string,
+        gitlabRemoteUrl?: string,
+        branch?: string,
+    }): GitlabRepoRef {
+        if (params.sha && !params.sha.match(GitShaRegExp.pattern)) {
+            throw new Error("You provided an invalid SHA: " + params.sha);
+        }
+        const result = new GitlabRepoRef(params.owner, params.repo, params.sha, params.rawApiBase, params.gitlabRemoteUrl, params.path);
+        result.branch = params.branch;
+        return result;
+    }
+
```

Comment:
GitHubRepoRef defines a `kind` property, which is nice https://github.com/atomist/automation-client/blob/87276736e56533ae235dd36fa7335906ca9ea91a/lib/operations/common/GitHubRepoRef.ts#L53
I don't think the BitBucketRepoRef does unfortunately. Maybe it should be added to RemoteRepoRef.